### PR TITLE
API to validate Quota activation rule

### PR DESCRIPTION
**Pull request by hsato03**
### Description

When entering an activation rule during the creation and editing of Quota tariffs, the defined rule is not validated. This way, errors are only identified while processing the tariff.

Because of this, the `quotaValidateActivationRule` API was created, which informs whether the activation rule is valid or not. Also, the activation rule field was added when creating and editing the tariff via UI.

#### API parameters

| Parameter | Description | Required |
| ------ | ------ |  ------ |
| activationrule | The activation rule that will be validated. | Yes |
| quotaresourcetype | tariff usage type to validate the variables used in the _script_. | Yes |

#### Example
```
🐱 > quota validateactivationrule activationrule="if (account.name == 'admin') { 0 } else { 300 }" usagetype=6
{
  "validactivationrule": {
    "activationrule": "if (account.name == 'admin') { 0 } else { 300 }",
    "isvalid": true,
    "message": "The script has no syntax errors and all variables are compatible with the given usage type.",
    "quotatype": "VOLUME"
  }
}
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

invalid-activation-rule

valid-activation-rule

### How Has This Been Tested?

| N° | Test | Valid activation rule |
| ------ | ------ | ------ |
| 1 | Activation rule with valid JS code and with variables compatible with the given usage type | Yes |
| 2 | Activation rule with valid JS code and with variables that are not compatible with the given usage type | No |
| 3 | Activation rule with invalid JS code | No |

### CODE REVIEW

**Comment by hsato03** [comment]
@blueorangutan package

**Review by bernardodemarco — APPROVED** [review]
LGTM, manually tested in a local environment.

## Test descriptions

Firstly, I validated whether the `quotaValidateActivationRule` API would correctly identify JavaScript syntax errors.

image

The API correctly recognized the syntax error, which is the expected behavior.

---

Next, I inserted a valid syntax, and no errors were returned in the API response.

image

---

Relating to the validations of the preset variables, I selected the `Template` usage type and inserted the `value.computingResources.cpuNumber` variable into the script. Since it is not compatible with the selected usage type, an error was returned.

image

---

Then, I changed the usage type to `Running VM` and the script had been successfully validated, since the variable `value.computingResources.cpuNumber` is compatible with the selected usage type.

image

File: server/src/main/java/org/apache/cloudstack/jsinterpreter/JsInterpreterHelper.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+        String fieldName = searchIntoObjectNodes(node);
+
+        if (fieldName == null) {
+            String key = fields.getKey();
+            if ("type".equals(key) && "CallExpression".equals(node.textValue())) {
+                callExpressions++;
+            }
+
+            if ("name".equals(key) || "property".equals(key)) {
```
Comment (BryanMLima) [inline server/src/main/java/org/apache/cloudstack/jsinterpreter/JsInterpreterHelper.java:117]: You could extract the Strings constants to variable, as they are used in other methods of this class.

**Review by BryanMLima — APPROVED** [review]
Just some minor suggestions, overall, LGTM.

**Comment by BryanMLima** [comment]
@hsato03, the build is failing due to the logger variable in the `VmwareHelper.java` class, it should be fixed by #9714.

File: plugins/database/quota/src/main/java/org/apache/cloudstack/api/response/QuotaValidateActivationRuleResponse.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+package org.apache.cloudstack.api.response;
+
+import com.cloud.serializer.Param;
+import com.google.gson.annotations.SerializedName;
+import org.apache.cloudstack.api.BaseResponse;
+
+public class QuotaValidateActivationRuleResponse extends BaseResponse {
+
+    @SerializedName("activationrule")
+    @Param(description = "The validated activation rule.")
```
[the remark this answers is not in the document]
Comment (hsato03) [reply plugins/database/quota/src/main/java/org/apache/cloudstack/api/response/QuotaValidateActivationRuleResponse.java:26]: At this point, I think the rule has already been validated.

**Comment by hsato03** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by hsato03** [comment]
@DaanHoogland the test error logs don't seem to be related to the PR. 
```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/usr/local/lib/python3.6/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/usr/local/lib/python3.6/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/usr/local/lib/python3.6/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/marvin/tests/smoke/test_usage.py", line 1025, in setUpClass
    cls.iso.id
Exception: ISO download failed exception: Failed to download ISO: 2785331c-21e7-47d5-91d5-ffb46b2dbe32
```
Could you run the smoke tests again?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by JoaoJandre** [comment]
Merging based on approvals and manual tests by @bernardodemarco. CI failures are not related to the PR (https://github.com/apache/cloudstack/pull/9605#issuecomment-2394982847)

**Pull request merged**